Building a React.js Application Using Django REST framework (#1475)
section-engineering-education/engineering-education: articles/django-react-app/index.md

from django.db import models
#create your models here.

# add this
class Todo(models.Model):
   title = models.CharField(max_length=100)
   description = models.TextField()
   completed = models.BooleanField(default=False)

   def _str_(self):
     return self.title
```

The model contains:

- Title: What the task is.

- Description: Give more explanation about a particular task.

- Completed: Completed is the status of a task; either completed or not completed.

Let us run migrations to add our model to the database schema.

```bash
python manage.py makemigrations todo
python manage.py migrate todo
```

We can add models to our Admin page using the `admin.site.register()` functions. In the todo app's `admin.py`, let's add the model to our admin page.

```python
# todo/admin.py

from django.contrib import admin
from .models import Todo # add this

class TodoAdmin(admin.ModelAdmin):
  list = ('title', 'description', 'completed')

  # Register your models here
  admin.site.register(Todo, TodoAdmin)
```

An administration page is crucial to any website that deals with dynamic content. Django was built to make that functionality easier.

```bash
python manage.py createsuperuser

> ```

Reviewer: Say, Let's create a superuser for the admin page to login.

Now, Let's create a serializer file.

Serializers allow complex data such as querysets and model instances to be converted to native python dataypes that can then be easily rendered into JSON, XML, or other content types.

```bash
touch todo/serializers.py
```

Let us add this to `serializers.py` file:

```python
# todo/serializers.py

from rest_framework import serializers
from .models import Todo

class TodoSerializer(serializers.ModelSerializer):
    class Meta:
        model = Todo
        fields = ('id' ,'title', 'description', 'completed')
```

Let's also update the `todo/views.py`:

```python
# todo/views.py

from django.shortcuts import render
from rest_framework import viewsets      # add this
from .serializers import TodoSerializer  # add this
from .models import Todo                 # add this

# Create your views here.

class TodoView(viewsets.ModelViewSet):   # add this
    serializer_class = TodoSerializer    # add this
    queryset = Todo.objects.all()        # add this

> ```

Reviewer: Explain what views are and give a brief explanation about the code snippet.

In the `backend/urls.py` we define the URL routes for the API:

```python
# backend/urls.py

from django.contrib import admin
from django.urls import path,include               # add this
from rest_framework import routers                 # add this
from todo import views                             # add this

router = routers.DefaultRouter()                   # add this
router.register(r'todos', views.TodoView, 'todo')  # add this

urlpatterns = [
    path('admin/', admin.site.urls),
    path('api/', include(router.urls))             # add this
]

> ```

Reviewer: Explain what URLs are and give a brief explanation about the code snippet.

* `todos/id` - Returns a specific Todo using the `id` primary key.

```
python manage.py runserver
```

![api](json.jpg)

We have set our backend let us move forward to frontend.

### Step 3: Frontend using React

To install `create-react-app` use the following command. `-g` stands for global as we are first installing `create-react-app` globally:

```bash
npm install -g create-react-app
```

While in the parent directory - `django-react-todoApp`- create a React application, `frontend`:

```bash
create-react-app frontend
```

To start server:

```bash
cd frontend
npm start
```

You should be able to see the default React app by now.

```javascript
// frontend/src/index.js

import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import reportWebVitals from './reportWebVitals';

ReactDOM.render(
    <App />,
    document.getElementById('root')
);

// If you want to start measuring performance in your app, pass a function
// to log results (for example: reportWebVitals(console.log))
// or send to an analytics endpoint. Learn more: https://bit.ly/CRA-vitals
reportWebVitals();
```

> `ReactDOM.render()` uses a DOM node in HTML to replace it with Javascript XML (JSX).

Reviewer: ```suggestion
`ReactDOM.render()` renders a React element into the DOM in the supplied container.
```

It enable us to integrate React in every foreign application.

> `ReactDOM.render()` takes two arguments. First is the JSX being rendered. Second argument specifies where React application hooks into HTML. 

Reviewer: ```suggestion
`ReactDOM.render()` takes two arguments. First, is the JSX being rendered. The second argument specifies the container element in the HTML page.
```

It expects an element with an `id='root'`.

Substitute the below code in `src/App.js`:

```javascript 
// frontend/src/App.js

import React, { Component } from "react";

const todoItems = [
  {
    id: 1,
    title: "Go to the market",
    description: "Buy ingredients to buy supper",
    completed: true
  },

  {
    id: 2,
    title: "Study",
    description: "Read algebra and history books about the coming test",
    completed: true
  },

  {
    id: 3,
    title: "Sally's Book",
    description: "Go to the library to rent Sally's books",
    completed: true
  },

];
class App extends Component {
  constructor(props) {
    super(props);
    this.state = {todoItems};
    };
    render() {
      return (
        <div>
          {this.state.todoItems.map(item => (
            <div>
              <h1>{item.title}</h1>
              <span>{item.description}</span>
            </div>
          ))}
        </div>
      )
    }
  }
export default App;
```

> We start by rendering a list of items. The list will be artificial data but later we will fetch the data from the API we created in the previous chapter.

Reviewer: ```suggestion
We start by rendering a list of items. The list will be artificial data but later we will fetch the data from the API we created in the previous steps.
```

> We define the list of items, an item in the list has an `id`, `title`, `description` and status of whether the task is completed or not, `completed`.

Reviewer: ```suggestion
We define a list of items. Each item has an `id`, `title`, `description`, and status of whether the task is completed or not, `completed`.
```

> We introduce the class constructor where we set the initial internal component state. In our case the internal state is the dummy list of items, `todoItems`.

Reviewer: ```suggestion
We introduce the class constructor where we set the initial state. In our case, the internal state is the dummy list of items, `todoItems`.
```

> We use the built-in JavaScript `map` functionality in our JavaScript XML (JSX). It allow us to iterate over the list of items and display them. We use the curly braces to encapsulate our item property since it evaluate Javascript expressions during compilation.

Reviewer: ```suggestion
We use the built-in JavaScript `map` functionality in our JavaScript XML (JSX). It allows us to iterate over the list of items and display them. We use curly braces to evaluate Javascript expressions during compilation in JSX.
```

When the component file is called it calls the `render()` method by default displaying th JSX syntax.

We use [arrow functions](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Functions/Arrow_functions) because they shorten our functions declaration.

Your User Interface should resemble the one below:

![frontend](dummy_data.jpg)

It is time to consume the API we created earlier.

```bash
cd backend
python manage.py runserver
```

> We will need to modify the `frontend/package.json`:

Reviewer: Please state what you are modifying and why.

```bash
npm install --save react-bootstrap-validation
```

We then import the `Jumbotron` in our `App.js` file.

To consume our API instead of the artificial data, update the `frontend/src/App.js` with the snippet below:

```javascript
// frontend/src/App.js

import React, { Component } from "react"
import Jumbotron from 'react-bootstrap/Jumbotron' // Add this

class App extends Component {
    state = {
      todoList: []
    };

    async componentDidMount() {
      try {
        const res = await fetch('http://localhost:8000/api/todos/');
        const todoList = await res.json();
        this.setState({
          todoList
        });
      } catch (e) {
        console.log(e);
    }
    }

    render() {
      return (
        <div>
          <Jumbotron>
          <h1>Todo App</h1>
          </Jumbotron>
          {this.state.todoList.map(item => (
            <div>
              <h1>{item.title}</h1>
              <span>{item.description}</span>
            </div>
          ))}
        </div>
      )
    }
  }
  
export default App;
```

We assign our `todoList` an empty array because we are going to fetch our data from an API.

> First, we wrap Fetch in a `try/catch` block to handle any network errors. We then call `fetch()` with the `await` keyword, where we pass our apiEndpoints. We're telling the  `async` function to stop executing untill the promise is resolved at which point it can resume execution and return the resolved value. Rather than getting promises, we will get back the parsed JSON data that we expect.

Reviewer: ```suggestion
First, we wrap `fetch()` in a `try/catch` block to handle any network errors. We then call `fetch()` with the `await` keyword, where we pass our API endpoints. We're telling the `async` function to stop executing until the promise is resolved at which point it'll resume execution and return the resolved value. Rather than getting promises, we will get back the parsed JSON data that we expect.
```

> We are using the `async` to fetch data from the API we created. Our application uses the `componentDidMount()` method from `React.Component` but defining it as an `async` function. This allows our use of `await` for each fetch. Using `await` outside of the `async` function results in a syntax error.

Reviewer: ```suggestion
Our application uses the `componentDidMount()` method from `React.Component` but we'll define it as an `async` function. This allows our use of `await` for each fetch. Using `await` outside of the `async` function results in a syntax error.
```

> `componentDidMount()` function is called by React to either fetch data from an External API or perform unique operations which need the JSX elements. In our case we use it for the former. Read more about [life cycle methods in React.](https://reactjs.org/docs/react-component.html)

Reviewer: ```suggestion
`componentDidMount()` function is called by React when a component is mounted for the first time. Read more about [life cycle methods in React.](https://reactjs.org/docs/react-component.html)
```

    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'todo', # Add this
]
```

Let's create a model to specify the Todo item fields. Modify `todo/models.py` as follows:

```python
# todo/models.py
from django.db import models
#create your models here.

# add this
class Todo(models.Model):
   title = models.CharField(max_length=100)
   description = models.TextField()
   completed = models.BooleanField(default=False)

   def _str_(self):
     return self.title
```

The model contains:

- Title: What the task is.

- Description: Give more explanation about a particular task.

- Completed: Completed is the status of a task; either completed or not completed.

We need to tell Django which attribute to use by default when it displays information about Todo. Django calls a `_str_()` method to display a simple representation of a model. Here we've defined a `_str_()` method that returns the string stored in the `title` attribute.

Let us run migrations to add our model to the database schema.

```bash
python manage.py makemigrations todo

> python manage.py migrate todo

Author: I'm on it still @zolomohan

from todo import views                             # add this

router = routers.DefaultRouter()                   # add this
router.register(r'todos', views.TodoView, 'todo')  # add this

urlpatterns = [
    path('admin/', admin.site.urls),
    path('api/', include(router.urls))             # add this
]
```

The next module is the `router.urls` which provides routing for our API.

The `router` enables  us to create the subsequent operations:

* `/todos/` - This returns a list of all the Todo items (Create and Read operations can be done here).

* `todos/id` - Returns a specific Todo using the `id` primary key.

```
python manage.py runserver
```

![api](json.jpg)

We have set our backend to let us move forward to the frontend.

### Step 3: Frontend using React

To install `create-react-app` use the following command. `-g` stands for global as we are first installing `create-react-app` globally:

```bash
npm install -g create-react-app
```

While in the parent directory - `django-react-todoApp`- create a React application, `frontend`:

```bash
create-react-app frontend
```

To start server:

```bash
cd frontend
npm start
```

You should be able to see the default React app by now.

> Next, let's install `bootstrap` and `reactstrap` to provide user interface tools.
> 
> ```bash
> npm install bootstrap@4.6.0 reactstrap@8.9.0 --legacy-peer-deps

Reviewer: ```suggestion
Next, let's install `bootstrap` and `reactstrap` to style the user interface.

```bash
npm install bootstrap@4.6.0 reactstrap@8.9.0 --legacy-peer-deps
```

Reviewer: Provide the link to Bootstrap and Reactstrap's homepage.

from .models import Todo

class TodoSerializer(serializers.ModelSerializer):
    class Meta:
        model = Todo
        fields = ('id' ,'title', 'description', 'completed')
```

We also update our `todo/views.py`:

```python
# todo/views.py

from django.shortcuts import render
from rest_framework import viewsets      # add this
from .serializers import TodoSerializer  # add this
from .models import Todo                 # add this

# Create your views here.

class TodoView(viewsets.ModelViewSet):   # add this
    serializer_class = TodoSerializer    # add this
    queryset = Todo.objects.all()        # add this

```

In the `backend/urls.py` we define the URL route for the API

```python
# backend/urls.py

from django.contrib import admin
from django.urls import path,include               # add this
from rest_framework import routers                 # add this
from todo import views                             # add this

router = routers.DefaultRouter()                   # add this
router.register(r'todos', views.TodoView, 'todo')  # add this

urlpatterns = [
    path('admin/', admin.site.urls),
    path('api/', include(router.urls))             # add this
]
```

> That completes the building of the API, The router class enables  us to create the subsequent operations:

Reviewer: ```suggestion
That completes the building of the API. The `router` class enables us to create the subsequent operations:
```

> - `/todos/` - This returns an inventory of all the Todo items(Create and Read operations may be done here).

Reviewer: ```suggestion
- `/todos/` - This returns a list of all the Todo items (Create and Read operations can be done here).
```

> - `todos/id` - Returns one Todo items by the id primary key.

Reviewer: ```suggestion
- `todos/id` - Returns a specific todo using the `id` primary key.
```

```
$ python manage.py runserver
```

![api](json.jpg)

We have set our backend let us move forward to frontend.

### Step 3: Frontend using React

> To install `create-react-app` use the following command, `-g` stands for global as we are first installing `create-react-app` globally:

Reviewer: ```suggestion
To install `create-react-app` use the following command. `-g` stands for global as we are first installing `create-react-app` globally:
```

```
$ npm install -g create-react-app
```

While in the parent directory - `django-react-todoApp`- create a React application, `frontend`:

```
$ create-react-app frontend
```

To start server:

```
$ cd frontend
$ npm start
```

You should be able to see the default React app by now.

> Add `bootstrap` and `reactstrap` for elegant UI:

Reviewer: ```suggestion
Install `bootstrap` and `reactstrap` to help us style the UI:
```

```
$ npm add bootstrap reactstrap
```

Head to `src/index.css` file and change the css to the one below:

```css
/_frontend/src/index.css _/

body {
  margin: 0;
  padding: 0;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Oxygen", "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue", sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  background-color: #282c34; }

.todo-title {
  cursor: pointer; }

.completed-todo {
  text-decoration: line-through; }

.tab-list > span {
  padding: 5px 8px;
  border: 1px solid #282c34;
  border-radius: 10px;
  margin-right: 5px;
  cursor: pointer; }

.tab-list > span.active {
  background-color: #282c34;
  color: #ffffff; }
```

> Import Bootstrap in `src/index.js`:

Reviewer: ```suggestion
Import the bootstrap file in `src/index.js`:
```

```javascript
// frontend/src/index.js

import React from 'react';
import ReactDOM from 'react-dom';
import 'bootstrap/dist/css/bootstrap.min.css';   // add this
import './index.css';
import App from './App';
import reportWebVitals from './reportWebVitals';

ReactDOM.render(<App />,document.getElementById('root'));

// If you want to start measuring performance in your app, pass a function
// to log results (for example: reportWebVitals(console.log))
// or send to an analytics endpoint. Learn more: https://bit.ly/CRA-vitals
reportWebVitals();
```

> Substitute the code in `src/App.js`:

Reviewer: ```suggestion
Substitute the below code in `src/App.js`:
```

                                name="description"
                                value={this.state.activeItem.description}
                                onChange={this.handleChange}
                                placeholder="Enter Todo description"
                            />
                        </FormGroup>   
                        <FormGroup check>
                            <Label for="completed">
                                <Input
                                    type="checkbox"
                                    name="completed"
                                    checked={this.state.activeItem.completed}
                                    onChange={this.handleChange}
                                />
                                Completed
                            </Label>
                        </FormGroup>  
                    </Form>
                </ModalBody> 
                <ModalFooter>
                    <Button color="success" onClick={() => onSave(this.state.activeItem)}>
                    Save
                    </Button>  
                </ModalFooter>
            </Modal>               
        );
    }
}
```

> `CustomModal` class which encapsulates the Modal component it comes from the `reactstrap`. Three fields are also defined; `Title`,`Description`,`Completed`.

Reviewer: ```suggestion
`CustomModal` class which encapsulates the Modal component comes from the `reactstrap`. Three fields are also defined; `Title`,`Description`,`Completed`.
```

                </div>   
                {this.state.modal ? (
                    <Modal
                        activeItem={this.state.activeItem} 
                        toggle={this.toggle}
                        onSave={this.handleSubmit}
                    />
                ) : null} 
            </main>
        );
    }

}
```

The User Interface should be like below at this point:

![styled](styled.jpg)

It is time to consume the API we created earlier.

```
$ cd backend
$ python manage.py runserver
```

We will install `axios`. It help us retrieve data from external API in our case the backend.

```
$ npm add axios
```

We will need to modify the `frontend/package.json`:

```javascript
// frontend/package.json

[...]
"name": "frontend",
  "version": "0.1.0",
  "private": true,
  "proxy": "http://localhost:8000",   // Add this
  "dependencies": {
    "@testing-library/jest-dom": "^5.11.9",
    "@testing-library/react": "^11.2.3",
    "@testing-library/user-event": "^12.6.0",
    "bootstrap": "^4.5.3",
    "react": "^17.0.1",
    "react-dom": "^17.0.1",
    "react-scripts": "4.0.1",
    "reactstrap": "^8.8.1",
    "web-vitals": "^0.2.4"
  },
  [...]
```

> The `axios` channels the API requests to the port `http://localhost:8000` where the backend(Django) execute it, enabling frontend to serve the requests. 

Reviewer: ```suggestion
The `axios` channels the API requests to the port `http://localhost:8000` where the Django server is running, enabling the frontend to make the requests. 
```
